Categories block: use supports flag for alignment + wide alignment support

ZebulanStanphill opened the pull request:
## Description
Previously #10635.

This PR changes the code for the Categories block to use the `supports` flag. It also adds support for the `wide` alignment to the Categories block, which strangely was lacking support for it despite already supporting the `full` alignment.

## How has this been tested?
I opened a post with an existing Categories block and checked to make sure setting alignments worked and applied the right classes in the editor and front-end.

## Types of changes
- The block now uses `supports: { align: true }` to add block alignment support. Note that the `align` attribute is still defined in PHP, as this is necessary for the alignments to work on the front-end.
- Added support for the `wide` alignment. (The block already supported the `full` alignment.)

ZebulanStanphill committed:
Categories block: use supports align flag + alignwide support

ZebulanStanphill commented:
I made a mistake on the PR... will fix in a moment.

ZebulanStanphill commented:
Okay, I've fixed the issue. This PR is now ready for review. @chrisvanpatten

chrisvanpatten commented:
@ZebulanStanphill I know you're keen to get consistency between blocks but it'll be trickier to merge this because of how big the changes are, without e2e tests.

Is there a reason you can't just update the block to use `supports.align` while keeping everything else? It would be much easier to guarantee the potential for regressions is minimised.

I wish we had more time to explore greater consistency between the server rendered blocks but because we're under the gun in terms of timeline I think it'd be much safer to simply update the alignment without making other potentially breaking changes.

chrisvanpatten commented on packages/block-library/src/categories/index.php:
```diff
@@ -94,6 +94,24 @@ function register_block_core_categories() {
+			'attributes'      => array(
+				'align'             => array(
+					'type' => 'string',
+					'enum' => array( 'left', 'center', 'right', 'wide', 'full' ),
```
The JS and the PHP don't communicate about this right now, but one thing to note is that by specifying the `enum` here, but not explicitly listing out the alignments in the JS `supports.align`, it could be possible that new alignment options are added to the editor but aren't supported here. I think that's not a huge deal, but worth noting.

chrisvanpatten commented on packages/block-library/src/categories/edit.js:
```diff
@@ -181,15 +177,6 @@ class CategoriesEdit extends Component {
 		return (
 			<Fragment>
 				{ inspectorControls }
-				<BlockControls>
```
Yay for red! ❤️

chrisvanpatten commented on packages/block-library/src/categories/index.js:
```diff
@@ -19,35 +19,11 @@ export const settings = {
 
 	category: 'widgets',
 
-	attributes: {
```
To keep us on track here I'd really recommend preserving the definitions in JS — it keeps the change surface minimised as we careen toward code freezes.

ZebulanStanphill replied:
@chrisvanpatten If the definitions remain in JS but are removed from PHP, I can't stop invalid classes from being applied. Should I define the attributes in both JS and PHP?

chrisvanpatten replied:
The JS should ensure that only valid alignments are set to the `align` attribute. It doesn't need to be reinforced in PHP (or to be more specific, we aren't currently enforcing it on PHP in this block and if an 'invalid' alignment somehow made it through to PHP it isn't a site-breaking issue).

chrisvanpatten reviewed (changes requested):
Small suggestions, per my previous comment. The less we change here, the more we can ensure it's going to be consistent without e2e tests.

ZebulanStanphill commented:
@chrisvanpatten Addressed feedback. Note that, using the Code Editor, set the `align` attribute to an invalid value, and an invalid class will be applied on the front-end. To fix this would require moving the attributes definiton to PHP and defining the `align` attribute there with an `enum` property specifying the valid values.

However, since this can only really happen if the user is intentionally trying to mess up the alignment classes, and since (I think) this behavior also occurs on `master`, I guess it is alright. A future PR could move the attributes definition to PHP.

ZebulanStanphill commented:
Tests are still failing and I can't figure out why. :confused: It seems to have something to do with #10682.

chrisvanpatten commented:
@ZebulanStanphill Tests are failing in a bunch of PRs; I can restart Travis a little later once the problem is fixed.

ZebulanStanphill commented:
@chrisvanpatten Tests are no longer failing on this, #10706, or #10707. :slightly_smiling_face:

ZebulanStanphill commented:
Closing in favor of #13215, which has some tweaks based on feedback on #10758, as well as a more accurate branch title.
